go/engine/subteam_create.go

```diff
@@ -0,0 +1,245 @@
+	}
+}
+
+func (e *SubteamCreateEngine) Name() string {
+	return "SubteamCreate"
+}
+
+func (e *SubteamCreateEngine) Prereqs() Prereqs {
+	return Prereqs{
+		Device: true,
+	}
+}
+
+func (e *SubteamCreateEngine) RequiredUIs() []libkb.UIKind {
+	return []libkb.UIKind{
+		libkb.LogUIKind,
+		libkb.SecretUIKind,
+	}
+}
+
+func (e *SubteamCreateEngine) SubConsumers() []libkb.UIConsumer {
+	return []libkb.UIConsumer{}
+}
+
+func (e *SubteamCreateEngine) Run(ctx *Context) (err error) {
+	defer e.G().CTrace(ctx.GetNetContext(), "SubteamCreateEngine", func() error { return err })()
+
+	me, err := libkb.LoadMe(libkb.NewLoadUserArg(e.G()))
+	if err != nil {
+		return err
+	}
+
+	deviceSigningKey, err := e.G().ActiveDevice.SigningKey()
+	if err != nil {
+		return err
+	}
+
+	parentTeam, err := teams.Get(ctx.GetNetContext(), e.G(), e.parentName)
+	if err != nil {
+		return err
+	}
+
+	// Subteam creation involves two links, one in the parent team's chain, and
+	// one to start the new subteam chain. The start of the new subteam chain
+	// (type "team.subteam_head") is very similar to the "team.root" sig that
+	// starts a root team, and so making that link is very similar to what the
+	// CreateTeamEngine does.
+
+	newSubteamSig, err := e.generateNewSubteamSigForParentChain(ctx, me, deviceSigningKey, parentTeam)
+	if err != nil {
+		return err
+	}
+
+	subteamHeadSig, secretboxes, err := e.generateHeadSigForSubteamChain(ctx, me, deviceSigningKey, parentTeam)
+	if err != nil {
+		return err
+	}
+
+	payload := make(libkb.JSONPayload)
+	payload["sigs"] = []interface{}{newSubteamSig, subteamHeadSig}
+	payload["per_team_key"] = secretboxes
+
+	_, err = e.G().API.PostJSON(libkb.APIArg{
+		Endpoint:    "sig/multi",
+		SessionType: libkb.APISessionTypeREQUIRED,
+		JSONPayload: payload,
+	})
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (e *SubteamCreateEngine) generateNewSubteamSigForParentChain(ctx *Context, me *libkb.User, signingKey libkb.GenericKey, parentTeam *teams.TeamSigChainState) (item *libkb.SigMultiItem, err error) {
+	newSubteamSigBody, err := teams.NewSubteamSig(me, signingKey, parentTeam, e.subteamFullyQualifiedName, e.subteamID)
+	newSubteamSigJSON, err := newSubteamSigBody.Marshal()
+	if err != nil {
+		return
+	}
+
+	v2Sig, err := makeSigchainV2OuterSig(
+		signingKey,
+		libkb.LinkTypeNewSubteam,
+		parentTeam.GetLatestSeqno()+1,
+		newSubteamSigJSON,
+		parentTeam.GetLatestLinkID(),
+		false, /* hasRevokes */
+	)
+	if err != nil {
+		return
+	}
+
+	item = &libkb.SigMultiItem{
+		Sig:        v2Sig,
+		SigningKID: signingKey.GetKID(),
+		Type:       string(libkb.LinkTypeNewSubteam),
+		SigInner:   string(newSubteamSigJSON),
+		TeamID:     parentTeam.GetID(),
+	}
+	return
+}
+
+func (e *SubteamCreateEngine) generateHeadSigForSubteamChain(ctx *Context, me *libkb.User, signingKey libkb.GenericKey, parentTeam *teams.TeamSigChainState) (item *libkb.SigMultiItem, boxes *PerTeamSharedSecretBoxes, err error) {
```

Comment:
how similar is this to what happens for a Root team and can we reuse and existing code?